sourcegraph/learn
# Add code highlighting

### What should this PR do?
Resolves [DEVED-171](https://sourcegraph.atlassian.net/browse/DEVED-171) by replacing `rehype-prism` with `prism-react-renderer` and using it to do both language syntax highlighting , and highlighting for selected code (as `<mark>`s). We accomplish this by replacing existing code blocks with a `Highlighter` component, which allows us to specify our preferred language for general syntax highlighting, and the code we want to highlight for emphasis (if any).

Part of the work of this PR will be updating our existing code blocks to use the new `Highlighter` component. I will do that work as part of this PR, but will track it separately, in [DEVED-192](https://sourcegraph.atlassian.net/browse/DEVED-192)

### Why are we making this change?
- This will allow us to highlight code for learning purposes; for example, in cases where we want readers to swap out a value for another value (e.g. "Replace `your-domain-here` with your domain").
- It also allows us to reduce the number of packages and vulnerabilities we have, from 1022 pkg/ 21 vul -> 781 pkg/ 8 vul.

### What are the acceptance criteria? 
- It should be possible to highlight selected blocks of code by passing them to the `Highlighter` component as a `matcher` prop.
- Syntax highlighting should work for the language passed to the `Highlighter` component as the `language` prop.
- It should be possible to include other languages, beyond what `prism-react-renderer` offers [out of the box](https://github.com/FormidableLabs/prism-react-renderer/blob/master/src/vendor/prism/includeLangs.js). See the work in this PR on Java for an example of this.

### How should this PR be tested?
- Check out the branch, remove your existing `node_modules`, and install dependencies with `npm ci`.
- Check that code blocks are correctly highlighted in existing tutorials for the language specified.
- Check at least one tutorial that has a `matcher` prop and make sure the selected code is highlighted as a `<mark>` element.
- Check that the instructions on using the `Highlighter` component in the README are clear and correct.

## Pull request process

**Reviewers**:

1. Test functionality using the criteria above.
2. Offer tips for efficiency, feedback on best practices, and possible alternative approaches and things that may not have been considered.
3. For shorter, "quick" PRs, use your best judgement on #​2.
4. Use a collaborative approach and provide resources and/or context where appropriate.
5. Provide screenshots/grabs where appropriate to show findings during review.

**Reviewees**:

1. Prefer incremental and appropriately-scoped changes.
2. Leave a comment on things you want explicit feedback on.
3. Respond clearly to comments and questions.
